### PR TITLE
Rework user-profile-behavior to use d2l-fetch

### PR DESCRIPTION
**Pull request by ryantmer**
The initial goal behind this was to stop the duplicate requests that were happening, as d2l-fetch-dedupe prevents this. Unfortunately, since this behavior is basically "fetch a bunch of stuff in semi-sequence", switching how the fetching is done basically means rewriting the whole thing... On the plus side, the actual flow logic is now all in one place (generateUserRequest), whereas before it was kind of all over the file and quite hard to follow. Hopefully, less so now!

TODO
- [x] make sure this works with folio evidence
- [x] TESTS, so many tests, now can write tests to address #10

### CODE REVIEW

File: d2l-user-profile-behavior.html [hunk 1]
```diff
@@ -32,25 +41,11 @@
...
-		_themeUrl: '',
-		_createIronAjaxRequest: function(request, url, headers, onResponse) {
-			request = request || document.createElement('iron-ajax');
-			request.url = url;
-			request.headers = headers;
-			this.listen(request, 'iron-ajax-response', onResponse);
-			this.listen(request, 'iron-ajax-error', onResponse);
-			return request;
-		},
 		_checkDoneRequests: function() {
```
Comment (ryantmer) [inline d2l-user-profile-behavior.html]: Imma be honest... I don't actually fully understand what this does :/

File: d2l-user-profile-behavior.html [hunk 2]
```diff
@@ -86,63 +81,79 @@
...
 			) {
 				this._previousUserCall = { userUrl: this.userUrl, token: this.token };
-				this._headers = {
-					Authorization: 'Bearer ' + this.token,
-					Accept: 'application/vnd.siren+json'
-				};
-				this._userRequest = this._createIronAjaxRequest(this._userRequest, this.userUrl, this._headers, '_onUserResponse');
-				this._userRequest.generateRequest();
+
+				this._fetchUser()
```
Comment (ryantmer) [inline d2l-user-profile-behavior.html]: I personally feel this is fairly easy to follow - hopefully. The flow is:

- get user, then try to get an image from Folio. If successful, we're done. If this fails,
- get enrollments -> if only one enrollment, get organization -> get organization image and use that. If this fails,
- get root -> get institution -> get theme -> use theme colour. If this fails,
- use `inherit` for the background colour.

File: d2l-user-profile-behavior.html [hunk 3]
```diff
@@ -86,63 +81,79 @@
...
+					if (response.ok) {
+						return response.json();
 					}
+					return Promise.reject(response.status);
+				})
+				.then(window.D2L.Hypermedia.Siren.Parse);
+		},
+		_fetchUser: function() {
+			return this._fetchSirenEntity(this.userUrl).then(function(userEntity) {
+				this._rootUrl = (userEntity.getLinkByRel(this.HypermediaRels.root) || {}).href;
```
Comment (ryantmer) [inline d2l-user-profile-behavior.html]: root, enrollments, and folio are all "entrypoints" for one of the workflows, so we might as well just keep all of their URLs.

File: .eslintrc.yml [hunk 4]
```diff
@@ -1,6 +1,8 @@
-extends: brightspace
+extends:
+  brightspace
 env:
   browser: true
+  es6: true
```
Comment (omsmith) [inline .eslintrc.yml]: What ES6 are we using? If we aren't transpiring is ES6 allowed?

File: d2l-user-profile-behavior.html [hunk 5]
```diff
@@ -1,17 +1,26 @@
 <link rel="import" href="../polymer/polymer.html">
-<link rel="import" href="../iron-ajax/iron-ajax.html">
 <link rel="import" href="../d2l-organization-hm-behavior/d2l-organization-hm-behavior.html">
 <link rel="import" href="../d2l-hm-constants-behavior/d2l-hm-constants-behavior.html">
 
 <script src="https://s.brightspace.com/lib/siren-parser/6.0.0/siren-parser.js"></script>
+<script src="https://s.brightspace.com/lib/d2lfetch/1.0.0/d2lfetch.js"></script>
```
Comment (omsmith) [inline d2l-user-profile-behavior.html]: So if we push a new version of fetch we need to update these in all the places everywhere?

File: d2l-user-profile-behavior.html [hunk 6]
```diff
@@ -1,17 +1,26 @@
...
+<script src="https://s.brightspace.com/lib/d2lfetch-auth/0.3.0/d2lfetch-auth-framed.js"></script>
```
Comment (omsmith) [inline d2l-user-profile-behavior.html]: Does framed work if we're not framed?

**Comment by apalaniuk** [comment]
@ryantmer Whoa.. I had a branch for this too, for the exact same reasons, that I was just finishing up and writing tests for. Oh well.. you got your PR up first, so I'll review this!

File: .eslintrc.yml (hunk 4 above)
Comment (apalaniuk) [inline .eslintrc.yml]: In this case, seems to only use promises.

File: d2l-user-profile-behavior.html (hunk 6 above)
Comment (ryantmer) [inline d2l-user-profile-behavior.html]: I don't believe so, no - I suppose we could add a check for framed-ness to determine which version to grab, yeah

File: d2l-user-profile-behavior.html (hunk 5 above)
Comment (ryantmer) [inline d2l-user-profile-behavior.html]: Unfortunately yep - we can definitely consider better way(s) to distribute it, granted!

**Comment by ryantmer** [comment]
@apalaniuk Damn, sorry! Didn't know anybody else was in the area :( I came across the duplicate requests, and initially was just going to dedupe those, but it kinda snowballed...

File: d2l-user-profile-behavior.html (hunk 1 above)
Comment (apalaniuk) [inline d2l-user-profile-behavior.html]: It's used [here](https://github.com/Brightspace/user-tile/blob/da8d625d94da913ce6d5e407062c6f38ad7c18f0/d2l-user-tile-auto.html#L29) to determine when to fade in from a skeleton

File: d2l-user-profile-behavior.html [hunk 7]
```diff
@@ -86,63 +81,79 @@
...
-					this._enrollmentsUrl = (userResponse.getLinkByRel(this.HypermediaRels.myEnrollments) || {}).href;
-					this._folioUrl = (userResponse.getLinkByRel(this.HypermediaRels.Folio.folio) || {}).href;
-					if (this._folioUrl) {
-						this._folioReqeust = this._createIronAjaxRequest(this._folioReqeust, this._folioUrl, this._headers, '_onFolioResponse');
-						this._folioReqeust.generateRequest();
+					if (image.class.indexOf('default-image') > -1) {
+						this._iconUrl = null;
 					} else {
-						this._makeEnrollmentsRequest();
+						this._iconUrl = (image.getLinkByRel(this.HypermediaRels.thumbnailAlternative) || {}).href;
```
Comment (apalaniuk) [inline d2l-user-profile-behavior.html]: I know you didn't introduce this, but is `thumbnailAlternative` even a thing? [Seems like it should be `thumbnailRegular` or `thumbnailSmall`](https://github.com/Brightspace/d2l-hm-constants-behavior/blob/master/d2l-hm-constants-behavior.html#L26)

File: d2l-user-profile-behavior.html (hunk 1 above)
Comment (ryantmer) [inline d2l-user-profile-behavior.html]: ohhh, gotcha. Thanks!

File: d2l-user-profile-behavior.html (hunk 7 above)
Comment (ryantmer) [inline d2l-user-profile-behavior.html]: So it should! I shall fix while I'm in here.

File: d2l-user-profile-behavior.html [hunk 8]
```diff
@@ -1,17 +1,26 @@
...
 
 <script src="https://s.brightspace.com/lib/siren-parser/6.0.0/siren-parser.js"></script>
+<script src="https://s.brightspace.com/lib/d2lfetch/1.0.0/d2lfetch.js"></script>
+<script src="https://s.brightspace.com/lib/d2lfetch-auth/0.3.0/d2lfetch-auth-framed.js"></script>
+<script src="https://s.brightspace.com/lib/d2lfetch-dedupe/0.6.0/d2lfetch-dedupe.js"></script>
 <script>
 	'use strict';
 
+	window.d2lfetch.use({name: 'auth', fn: window.d2lfetch.auth});
+	window.d2lfetch.use({name: 'dedupe', fn: window.d2lfetch.dedupe});
```
Comment (CodeBaboon) [inline d2l-user-profile-behavior.html]: I'm not sure these should be here, or maybe we need a better way of doing this, but if we `use` these same middlewares at the parent-portal-ui app level then I think we will double them up here. My original thinking behind the middleware was the 'app' level project would be the thing to set up what it felt was appropriate since it has all the context and lower-level components would simply use fetch straight up. That removes needing to decide if you are framed or not for auth, whether you want to dedupe, what cache strategy, etc.

File: d2l-user-profile-behavior.html [hunk 9]
```diff
@@ -86,63 +81,79 @@
...
+				})
+			});
+
+			return window.d2lfetch
+				.fetch(request)
+				.then(function(response) {
+					if (response.ok) {
+						return response.json();
 					}
+					return Promise.reject(response.status);
```
Comment (omsmith) [inline d2l-user-profile-behavior.html]: Far preferable to reject `Error`s.

File: d2l-user-profile-behavior.html [hunk 10]
```diff
@@ -86,63 +81,79 @@
...
+				this._enrollmentsUrl = (userEntity.getLinkByRel(this.HypermediaRels.myEnrollments) || {}).href;
+				this._folioUrl = (userEntity.getLinkByRel(this.HypermediaRels.Folio.folio) || {}).href;
+
+				var displayNameEntity = userEntity.getSubEntityByRel(this.HypermediaRels.displayName);
+				if (displayNameEntity) {
+					this._name = displayNameEntity.properties && displayNameEntity.properties.name;
 				}
 
-				this._rootUrl = (userResponse.getLinkByRel(this.HypermediaRels.root) || {}).href;
+				var profileEntity = userEntity.getSubEntityByRel(this.HypermediaRels.profileImage);
```
Comment (omsmith) [inline d2l-user-profile-behavior.html]: `userProfile`

File: d2l-user-profile-behavior.html [hunk 11]
```diff
@@ -86,63 +81,79 @@
...
+				if (profileEntity) {
+					var image = profileEntity.getSubEntityByRel(this.HypermediaRels.profileImage);
 
-				if (this.options.background) {
-					this._enrollmentsUrl = (userResponse.getLinkByRel(this.HypermediaRels.myEnrollments) || {}).href;
-					this._folioUrl = (userResponse.getLinkByRel(this.HypermediaRels.Folio.folio) || {}).href;
-					if (this._folioUrl) {
-						this._folioReqeust = this._createIronAjaxRequest(this._folioReqeust, this._folioUrl, this._headers, '_onFolioResponse');
-						this._folioReqeust.generateRequest();
+					if (image.class.indexOf('default-image') > -1) {
```
Comment (omsmith) [inline d2l-user-profile-behavior.html]: Would personally prefer `!==`

File: d2l-user-profile-behavior.html (hunk 7 above)
Comment (jstefaniuk-d2l) [inline d2l-user-profile-behavior.html]: There was a reason for using this one that I can't remember for the life of me

File: d2l-user-profile-behavior.html (hunk 8 above)
Comment (ryantmer) [inline d2l-user-profile-behavior.html]: Works for me! I'll remove it and note it in the readme that it should be set up by at least the calling component.

File: d2l-user-profile-behavior.html (hunk 7 above)
Comment (apalaniuk) [inline d2l-user-profile-behavior.html]: As in, there's a reason for us to keep it that way? Or that used to be a thing (though I can't seem to
 find it)? 😕

Comment (ryantmer) [inline d2l-user-profile-behavior.html]: My guess is that the `alternative` links can be used to point to the actual resource (i.e. the image itself) rather the resource entity. Buuuut if the rel in question doesn't exist, then yeah, no going to do much anyway :/

Comment (jstefaniuk-d2l) [inline d2l-user-profile-behavior.html]: Yeah I think it was originally needed so that parents impersonating their children could see the profile pics without the server sending unauthorized errors, I guess the rel just got lost at some point or something?

File: d2l-user-profile-behavior.html [hunk 12]
```diff
@@ -153,7 +153,7 @@
 				for (var i = 0; i < tiles.length; i++) {
 					var content = tiles[i].getSubEntityByRel(this.HypermediaRels.Folio.contentItem);
 					var type = content.properties.type;
-					if (type === 'Jpg' && type !== 'Png') {
+					if (type === 'Jpg' || type === 'Png') {
```
Comment (jstefaniuk-d2l) [inline d2l-user-profile-behavior.html]: mind also adding Gif for equality to the user tile? https://github.com/Brightspace/parent-portal-ui/blob/master/src/widgets/d2l-parent-portal-folio-widget/d2l-evidence-tile-wrapper.html#L105

Comment (apalaniuk) [inline d2l-user-profile-behavior.html]: Was going to play it safe, but seeing that case, I'm good with putting it in. Thanks! 👍

**Comment by apalaniuk** [comment]
Code LGTM. Test cases (as consumed in Brightspace for Parents):

- [ ] Student has folio evidence shared with parent. Background of student's user tile is set to evidence content.
- [ ] Student has no folio evidence shared with parent. Has one enrollment, and organization has image. Org image used for student user tile background.
- [ ] Student has more than one enrollment. Institution theme used.
- [ ] Organization does not have image. Institution theme used.
- [ ] Organization does not have image. No institution theme set. Background color set to 'initial'.
